### PR TITLE
Refactor `PrimitiveTypeTable` for Clippy

### PR DESCRIPTION
**Pull request by camsteffen**
I removed `PrimitiveTypeTable` and added `PrimTy::ALL` and `PrimTy::from_name` in its place. This allows Clippy to use `PrimTy::from_name` for the `builtin_type_shadow` lint, and a `const` list of primitive types is deleted from Clippy code (the goal). All changes should be a little faster, if anything.

### CODE REVIEW

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -2057,6 +2057,26 @@ pub enum PrimTy {
 }
 
 impl PrimTy {
+    pub const ALL: [Self; 17] = [
```
Comment (oli-obk) [inline compiler/rustc_hir/src/hir.rs:2060]: Please leave a comment on this mentioning that any changes should also be reflected in `from_name`. Also leave the opposite comment on `from_name`.

**Review by oli-obk — CHANGES_REQUESTED** [review]
Great refactoring!

Could you also add some documentation on the new constant and function? Other than those nits, r=me

**Comment by camsteffen** [comment]
Good! I wasn't sure if the associated const was kosher. 😄 

Changes made.

**Pull request merged**